File: ClassFiles/week6/edit-fix.r
library(stringr)

#' Computes the edit distance between two strings
#'
#' @param s1 -- a string
#' @param s2 -- a string
#'
#' @return the edit distance between the strings, 
#'         a non-negative integer
#' 
edit_distance <- function(s1, s2) {
    str1 <- unlist(strsplit(s1, ""))
    len1 <- length(str1)

    str2 <- unlist(strsplit(s2, ""))
    len2 <- length(str2)

    # The edit matrix naturally indexed from 0:len1, 0:len2,
    # so in R, we add 1 to the actual index to match.
    # Initialize to NA to make errors clearer.
    edit <- matrix(NA, len1+1, len2+1)

    # Q: What do these initializations mean?
    edit[,1] <- 0:len1
    edit[1,] <- 0:len2

    for( index1 in 2:(len1+1) ) {
        for( index2 in 2:(len2+1) ) {
            #ATTN:FIX
            edit[index1, index2] <- NA #ATTN:FIX
        }
    }

    return( edit[len1+1, len2+1])
}                     


gaps <- function(gap_length) {
    return( paste(rep("_", gap_length), collapse="") )
}

#' Computes the alignment and edit distance between two strings
#'
#' @param s1 -- a string
#' @param s2 -- a string
#'
#' @return a list with two attributes:
#'           - distance  :: the edit distance between the strings,
#'           - alignment :: a pair of strings with '_' characters placed between them
#' 
edit_alignment <- function(s1, s2) {
    str1 <- unlist(strsplit(s1, ""))
    len1 <- length(str1)

    str2 <- unlist(strsplit(s2, ""))
    len2 <- length(str2)

    # The edit matrix naturally indexed from 0:len1, 0:len2,
    # so in R, we add 1 to the actual index to match.
    # Initialize to NA to make errors clearer.
    edit <- matrix(NA, len1+1, len2+1)
    align <- array("", c(len1+1, len2+1, 2))

    # Q: What do these initializations mean?
    edit[,1] <- 0:len1
    edit[1,] <- 0:len2

    align[1,1,] <- ""
    for ( i in 2:(len1+1) ) {
        align[i, 1, 1] <- "NA" #ATTN:FIX  (hint: you can use str_sub or substring)
        align[i, 1, 2] <- gaps(i-1)
    }
    for ( j in 2:(len2+1) ) {
        align[1, j, 1] <- "NA" #ATTN:FIX
        align[1, j, 2] <- "NA" #ATTN:FIX
    }

    for( index1 in 2:(len1+1) ) {
        for( index2 in 2:(len2+1) ) {
            #ATTN:FIX

            edit[index1, index2] <- NA #ATTN:FIX

            if ( TRUE ) { #ATTN:FIX
                align[index1, index2, 1] <- str_c("NA", "NA") #ATTN:FIX
                align[index1, index2, 2] <- str_c("NA", "NA") #ATTN:FIX
            } else if ( TRUE ) { #ATTN:FIX
                align[index1, index2, 1] <- str_c("NA", "NA") #ATTN:FIX
                align[index1, index2, 2] <- str_c("NA", "NA") #ATTN:FIX
            } else {
                align[index1, index2, 1] <- str_c("NA", "NA") #ATTN:FIX
                align[index1, index2, 2] <- str_c("NA", "NA") #ATTN:FIX
            }
        }
    }

    return( list(distance=edit[len1+1, len2+1], alignment=align[len1+1,len2+1,]) )
}                     

File: ClassFiles/week6/edit-fix.py
import numpy as np


def edit_distance(s1, s2):
    """Computes the edit distance between two strings

    Parameters s1 and s2 are strings.

    Returns the edit distance, a non-negative integer, between the strings.

    """
    len1 = len(s1)
    len2 = len(s2)
    edit = np.full((len1 + 1, len2 + 1), np.inf)

    edit[:, 0] = np.arange(len1 + 1)
    edit[0, :] = np.arange(len2 + 1)

    for index1 in range(1, len1 + 1):
        for index2 in range(1, len2 + 1):
            # ATTN:FIX
            edit[index1, index2] = np.inf  # ATTN:FIX
    edit = edit.astype(int)
    return edit[len1, len2]


def gaps(gap_length):
    return "_" * gap_length


def edit_alignment(s1, s2):
    """Computes the alignment and edit distance between two strings

    Parameters s1 and s2 are strings.

    Returns a 3-tuple containing the the edit distance, the first string
    with alignment characters _ included, and the second string with alignment
    characters included.

    """
    len1 = len(s1)
    len2 = len(s2)

    edit = np.full((len1 + 1, len2 + 1), np.inf)
    edit[:, 0] = np.arange(len1 + 1)
    edit[0, :] = np.arange(len2 + 1)

    align = np.full((len1 + 1, len2 + 1, 2), '', dtype=object)

    for i in range(1, len1 + 1):
        align[i, 0, 0] = "NA"  # ATTN:FIX
        align[i, 0, 1] = gaps(i)

    for j in range(1, len2 + 1):
        align[0, j, 0] = "NA"  # ATTN:FIX
        align[0, j, 1] = "NA"  # ATTN:FIX

    for index1 in range(1, len1 + 1):
        for index2 in range(1, len2 + 1):
            # ATTN:FIX
            edit[index1, index2] = np.inf  # ATTN:FIX

            if True:
                align[index1, index2, 0] = "NA"  # ATTN:FIX
                align[index1, index2, 1] = "NA"  # ATTN:FIX
            elif True:
                align[index1, index2, 0] = "NA"  # ATTN:FIX
                align[index1, index2, 1] = "NA"  # ATTN:FIX
            else:
                align[index1, index2, 0] = "NA"  # ATTN:FIX
                align[index1, index2, 1] = "NA"  # ATTN:FIX
    edit = edit.astype(int)
    return (edit[len1, len2], align[len1, len2, 0], align[len1, len2, 1])
